docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,1063 @@
+    approach to multitenancy, we should enforce are like singleton (a
+    provider, identified by Kind & name, should be unique at cluster level)
+    It will be nice to provide a nice UX around kubectl get <providers>
+1. What should the Group/Version for this API be?
+    Currently, we have it as `management.cluster.x-k8s.io/v1alpha1`. Should
+    the version be `v1alpha4` to match the CAPI version it is being introduced
+    in? This would be consistent with what we do for experiments.
+1. Should we have multiple types (i.e. CoreProvider, BootstrapProvider,
+   ControlPlaneProvider and InfrastructureProvider) vs. a single Provider
+   type?
+   Even though we can use CRD categories to group them, the output may
+   not be as readable as we’d want.
+
+#### Example API Usage
+
+1. As an admin, I want to install the aws infrastructure provider with
+   specific controller flags.
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+ name: aws-variables
+ namespace: capa-system
+type: Opaque
+data:
+ AWS_REGION: ...
+ AWS_ACCESS_KEY_ID: ...
+ AWS_SECRET_ACCESS_KEY: ...
+---
+apiVersion: management.cluster.x-k8s.io/v1alpha1
+kind: InfrastructureProvider
+metadata:
+ name: aws
+ namespace: capa-system
+spec:
+ version: v0.6.0
+ secretName: aws-variables
+ manager:
+   # These top level controller manager flags, supported by all the providers.
+   # These flags come with sensible defaults, thus requiring no or minimal
+   # changes for the most common scenarios.
+   metricsAddress: ":8181"
+   syncPeriod: 660
+ fetchConfig:
+   url: https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases
+ deployment:
+   containers:
+   - name: manager
+     args:
+         # These are controller flags that are specific to a provider; usage
+         # is reserved for advanced scenarios only.
+         awscluster-concurrency: 12
+         awsmachine-concurrency: 11
+```
+
+2. As an admin, I want to install aws infrastructure provider but override
+   the container image of the CAPA deployment.
+
+```yaml
+---
+apiVersion: management.cluster.x-k8s.io/v1alpha1
+kind: InfrastructureProvider
+metadata:
+ name: aws
+ namespace: capa-system
+spec:
+ version: v0.6.0
+ secretName: aws-variables
+ deployment:
+   containers:
+   - name: manager
+     image: gcr.io/myregistry/capa-controller:v0.6.0-foo
+```
+
+3. As an admin, I want to change the resource limits for the manager pod in
+   my control plane provider deployment.
+
+```yaml
+---
+apiVersion: management.cluster.x-k8s.io/v1alpha1
+kind: ControlPlaneProvider
+metadata:
+ name: kubeadm
+ namespace: capi-kubeadm-control-plane-system
+spec:
+ version: v0.3.10
+ secretName: capi-variables
+ deployment:
+   containers:
+   - name: manager
+     resources:
+       limits:
+         cpu: 100m
+         memory: 30Mi
+       requests:
+         cpu: 100m
+         memory: 20Mi
+```
+
+4. As an admin, I would like to fetch my azure provider components from a
```

Comment:
would be good to include an example of an infrastructure provider with no credentials as well

Reply:
I can add one in there but that would just be omitting the `secretName` field since it's optional anyways. But I'll look for an example, I don't think it hurts to have more examples. 🙂